File: promise/src/test/java/com/voxeet/testpromise/sub_promise/promise_io/PromiseTestWithSumDelayed2.java
package com.voxeet.testpromise.sub_promise.promise_io;

import com.voxeet.promise.Promise;
import com.voxeet.promise.solve.PromiseExec;
import com.voxeet.testpromise.utils.AndroidMockUtil;

import org.junit.Before;
import org.junit.Test;

import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;

/**
 * Created by kevinleperf on 06/04/2018.
 */

public class PromiseTestWithSumDelayed2 {

    @Before
    public void setHandler() {
        Promise.setHandler(AndroidMockUtil.mockMainThreadHandler());
    }

    @Test
    public void test() throws InterruptedException {
        final CountDownLatch latch = new CountDownLatch(1);
        final int[] final_result = {0};

        System.out.println("executing test");
        execute()
                .then((PromiseExec<Integer, Integer>) (result, solver) -> {
                    //testing here resolving from promise in a chain
                    solver.resolve(new Promise<>(solver1 -> solver1.resolve(result)));
                })
                .then((result, solver) -> final_result[0] = result)
                .error(error -> {
                    System.out.println("error catched");
                    final_result[0] = 0;
                    error.printStackTrace();
                    latch.countDown();
                });

        //6s are enough
        latch.await(6, TimeUnit.SECONDS);
        if (final_result[0] != 10) {
            throw new IllegalStateException("Expected 10... having " + final_result[0]);
        } else {
            System.out.println("having result " + final_result[0]);
        }
    }

    private Promise<Integer> execute() {
        return new Promise<>(solver -> solver.resolve(10));
    }
}

File: promise/src/test/kotlin/com/voxeet/testpromise/subpromise/promise_io/PromiseTestWithSumDelayed2.kt
package com.voxeet.testpromise.subpromise.promise_io

import com.voxeet.promise.HandlerFactory
import com.voxeet.promise.Promise
import com.voxeet.promise.solve.Solver
import com.voxeet.testpromise.mockedhandler
import org.junit.Before
import org.junit.Test
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit

class PromiseTestWithSumDelayed2 {
    @Before
    fun setHandler() {
        HandlerFactory.setHandler(mockedhandler())
    }

    @Test
    @Throws(InterruptedException::class)
    fun test() {
        val latch = CountDownLatch(1)
        val final_result = intArrayOf(0)
        println("executing test")
        execute()
            .then { result: Int?, solver: Solver<Int?> ->
                //testing here resolving from promise in a chain
                solver.resolve(Promise { solver1: Solver<Int?> ->
                    solver1.resolve(
                        result
                    )
                })
            }
            .then { final_result[0] = it!! }
            .error { error: Throwable ->
                println("error catched")
                final_result[0] = 0
                error.printStackTrace()
                latch.countDown()
            }

        //6s are enough
        latch.await(6, TimeUnit.SECONDS)
        check(final_result[0] == 10) { "Expected 10... having " + final_result[0] }
        println("having result " + final_result[0])
    }

    private fun execute(): Promise<Int?> {
        return Promise { solver: Solver<Int?> -> solver.resolve(10) }
    }
}
